test/k8sT/DatapathConfiguration.go

```diff
@@ -645,9 +645,86 @@ var _ = Describe("K8sDatapathConfig", func() {
+		SkipItIf(helpers.RunsWithKubeProxy, "With native routing", func() {
+			options := map[string]string{
+				"hostFirewall":         "true",
+				"tunnel":               "disabled",
+				"autoDirectNodeRoutes": "true",
+			}
+			// We can't rely on gke.enabled because it enables
+			// per-endpoint routes which are incompatible with
+			// the host firewall.
+			if helpers.RunsOnGKE() {
+				options["gke.enabled"] = "false"
+			}
+			deploymentManager.DeployCilium(options, DeployCiliumOptionsAndDNS)
+			testHostFirewall(kubectl)
+		})
 	})
 })
 
+func testHostFirewall(kubectl *helpers.Kubectl) {
+	randomNs := deploymentManager.DeployRandomNamespaceShared(DemoHostFirewall)
+	deploymentManager.WaitUntilReady()
+
+	demoHostPolicies := helpers.ManifestGet(kubectl.BasePath(), "host-policies.yaml")
+	By(fmt.Sprintf("Applying policies %s", demoHostPolicies))
+	_, err := kubectl.CiliumPolicyAction(randomNs, demoHostPolicies, helpers.KubectlApply, helpers.HelperTimeout)
+	ExpectWithOffset(1, err).Should(BeNil(), fmt.Sprintf("Error creating resource %s: %s", demoHostPolicies, err))
+
+	defer func() {
+		By(fmt.Sprintf("Removing policies %s", demoHostPolicies))
+		_, err = kubectl.CiliumPolicyAction(randomNs, demoHostPolicies, helpers.KubectlDelete, helpers.HelperTimeout)
+		ExpectWithOffset(1, err).Should(BeNil(), fmt.Sprintf("Error removing resource %s: %s", demoHostPolicies, err))
+	}()
+
+	By("Checking host policies on ingress from local pod")
+	testHostFirewallWithPath(kubectl, randomNs, "zgroup=testClient", "zgroup=testServerHost", false)
+
+	By("Checking host policies on ingress from remote pod")
+	testHostFirewallWithPath(kubectl, randomNs, "zgroup=testClient", "zgroup=testServerHost", true)
+
+	By("Checking host policies on egress to local pod")
+	testHostFirewallWithPath(kubectl, randomNs, "zgroup=testClientHost", "zgroup=testServer", false)
+
+	By("Checking host policies on egress to remote pod")
+	testHostFirewallWithPath(kubectl, randomNs, "zgroup=testClientHost", "zgroup=testServer", true)
+
+	By("Checking host policies on ingress from remote node")
+	testHostFirewallWithPath(kubectl, randomNs, "zgroup=testServerHost", "zgroup=testClientHost", true)
+
+	By("Checking host policies on egress to remote node")
+	testHostFirewallWithPath(kubectl, randomNs, "zgroup=testClientHost", "zgroup=testServerHost", true)
+}
+
+func testHostFirewallWithPath(kubectl *helpers.Kubectl, randomNs, client, server string, crossNodes bool) {
+	srcPod, srcPodJSON := fetchPodsWithOffset(kubectl, randomNs, "client", client, "", crossNodes, 3)
+	srcHost, err := srcPodJSON.Filter("{.status.hostIP}")
+	ExpectWithOffset(2, err).Should(BeNil(), "Failure to retrieve host of pod %s", srcPod)
+
+	dstPod, dstPodJSON := fetchPodsWithOffset(kubectl, randomNs, "server", server, srcHost.String(), crossNodes, 3)
+	podIP, err := dstPodJSON.Filter("{.status.podIP}")
+	ExpectWithOffset(2, err).Should(BeNil(), "Failure to retrieve IP of pod %s", dstPod)
```

Comment:
[Non blocking] It looks like the `fetch/filter` pattern is common in this file, would it make sense to refactor in a helper function?

Reply:
I was a bit hesitant to do that due to the large number of arguments `fetchPodsWithOffset` already takes. Such a wrapper would take 8 arguments and return 4 variables :-/  Maybe there's a better way to implement that in Go?

Reply:
Right... I guess the obvious solution would be to logically regroup arguments in wrapper state objects, e.g.
```go
helpers.fetchPodsOptions{
    Kubectl: kubectl,
    ...
}
```
and likewise for return values.

But I suppose this would only work if it's done at the highest possible level, and not here, with the state objects being copied/enriched/passed from higher levels to sub-levels. But this would be a bigger refactor 🤔